File: samples/uwp/SetOfCsModules/Module3.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Azure.IoT.Gateway;

namespace SetOfCsModules
{
    public sealed class Module3 : IGatewayModule
    {
        public void Create(MessageBus bus, string configuration)
        {
            System.Diagnostics.Debug.WriteLine("SetOfCsModules.Module3.Create");

            new Task(() => {
                for (int i = 0; i < 3; i++)
                {
                    new System.Threading.ManualResetEvent(false).WaitOne(1000);
                    var props = new Dictionary<string, string>();
                    props.Add("SetOfCsModules.Module3", i.ToString());
                    bus.Publish(new Microsoft.Azure.IoT.Gateway.Message("SetOfCsModules.Module3 says hello!!!", props));
                }
            }).Start();
        }

        public void Destroy()
        {
            System.Diagnostics.Debug.WriteLine("SetOfCsModules.Module3.Destroy");
        }

        public void Receive(Message received_message)
        {
            var contentBytes = received_message.GetContent();
            StringBuilder content = new StringBuilder();
            for (int i = 0; i < contentBytes.Count; i++)
            {
                char ch = (char)contentBytes[i];
                if (ch != '\0')
                {
                    content.Append(ch);
                }
            }

            var props = received_message.GetProperties();
            StringBuilder propsString = new StringBuilder();
            foreach (var key in props.Keys)
            {
                propsString.Append(key);
                propsString.Append("=");
                propsString.Append(props[key] as string);
                propsString.Append(", ");
            }

            StringBuilder sb = new StringBuilder();
            sb.Append("SetOfCsModules.Module3.Receive: Content=" + content.ToString() + "\r\n");
            sb.Append("SetOfCsModules.Module3.Receive: Properties={" + propsString.ToString() + "}\r\n");
            System.Diagnostics.Debug.Write(sb.ToString());
        }
    }
}

File: samples/uwp/SetOfCppModules/Module3.cpp
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

#include "pch.h"
#include "Module3.h"

using namespace SetOfCppModules;
using namespace Platform;
using namespace concurrency;

Module3::Module3()
{
}

void SetOfCppModules::Module3::Create(Microsoft::Azure::IoT::Gateway::MessageBus ^bus, Platform::String ^configuration)
{
	std::wstring formattedText = L"SetOfCppModules::Module3::Create\r\n";
	OutputDebugString(formattedText.c_str());

	create_async([bus]() {
		wchar_t count[3];
		for (int i = 0; i < 3; i++)
		{
			Sleep(1000);

			Platform::String^ content = "SetOfCppModules::Module3 says hello!!!";
			auto properties = ref new Platform::Collections::Map<Platform::String^, Platform::String^>();

			swprintf(count, 3, L"%d", i);
			properties->Insert("SetOfCppModules::Module3", ref new Platform::String(count));

			auto msg = ref new Microsoft::Azure::IoT::Gateway::Message(content, properties->GetView());
			bus->Publish(msg);
		}
	});
}

void SetOfCppModules::Module3::Destroy()
{
	std::wstring formattedText = L"SetOfCppModules::Module3::Destroy\r\n";
	OutputDebugString(formattedText.c_str());
}

void SetOfCppModules::Module3::Receive(Microsoft::Azure::IoT::Gateway::Message ^received_message)
{
	auto contentBytes = received_message->GetContent();
	std::wstring content;
	for (unsigned int i = 0; i < contentBytes->Size; i++)
	{
		content += (wchar_t)contentBytes->GetAt(i);
	}

	auto props = received_message->GetProperties();

	std::wstring propsString;
	for (auto iter = begin(props); iter != end(props); iter++)
	{
		propsString += (*iter)->Key->Data();
		propsString += L"=";
		propsString += (*iter)->Value->ToString()->Data();
		propsString += L", ";
	}

	std::wstring output = L"SetOfCppModules.Module3.Receive: Content=";
	output += content.c_str();
	output += L"\r\n";
	output += L"SetOfCppModules.Module3.Receive: Properties={";
	output += propsString.c_str();
	output += L"}\r\n";
	OutputDebugString(output.c_str());
}
